rasa/core/tracker_store.py

```diff
@@ -471,12 +471,49 @@ def save(self, tracker, timeout=None):
         if self.event_broker:
             self.stream_events(tracker)
 
-        state = tracker.current_state(EventVerbosity.ALL)
+        additional_events = self._additional_events(tracker)
 
         self.conversations.update_one(
-            {"sender_id": tracker.sender_id}, {"$set": state}, upsert=True
+            {"sender_id": tracker.sender_id},
+            {"$push": {"events": {"$each": [e.as_dict() for e in additional_events]}}},
+            upsert=True,
         )
 
+    def _additional_events(self, tracker: DialogueStateTracker) -> Iterator:
+        """Return events from the tracker which aren't currently stored.
+
+        Args:
+            tracker: Tracker to inspect.
+
+        Returns:
+            List of serialised events that aren't current stored.
+
+        """
+        stored = self.conversations.find_one({"sender_id": tracker.sender_id})
+        n_events = len(stored.get("events", [])) if stored else 0
```

Comment:
you could use `self.number_of_existing_events()`

Reply:
Although you save the de-serialization this way. Mhm, if you have an idea to de-duplicate things (we are doing the same thing in the event broker, SQLTracker store in slightly different flavours), then change it. Otherwise keep it

Reply:
exactly, I wanted to avoid having to re-fetch things. the way we calculate n_events is different in the two tracker stores, so I don't think it makes sense to share code between the two